sphinx/_cli/__init__.py

```diff
@@ -0,0 +1,179 @@
+"""Base 'sphinx' command.
+
+Subcommands are loaded lazily from the ``_COMMANDS`` table for performance.
+
+All subcommand modules must define three attributes:
+
+- ``parser_description``, a description of the subcommand. The first paragraph
+  is taken as the short description for the command.
+- ``set_up_parser``, a callable taking and returning an ``ArgumentParser``. This
+  function is responsible for adding options and arguments to the subcommand's
+  parser..
+- ``run``, a callable taking parsed arguments and returning an exit code. This
+  function is responsible for running the main body of the subcommand and
+  returning the exit status.
+
+The entire ``sphinx._cli`` namespace is private, only the command line interface
+has backwards-compatability guarantees.
+"""
+
+import argparse
+import locale
+import sys
+
+from sphinx._cli.console_utilities import color_terminal, nocolor
+from sphinx.locale import __, init_console
+
+if False:
+    # NoQA
+    from collections.abc import Callable, Iterator, Sequence
+    from typing import Dict, Tuple
```

Comment:
i'm not sure I understand the motivation behind this pattern. Do you have benchmarks that show this makes a significant different to import time (assuming that's what you're optimising for)?

It's not very IDE friendly

Reply:
(Responding to your comment https://github.com/sphinx-doc/sphinx/commit/0466fa88ebd7b97e7ddda292450a349f17dde4a7#commitcomment-85236051 here as well so as not to have a discussion on a commit) 

Any typing imports slow down import time by 4% on my computer, and I want the CLI to be as responsive as possible. (Ideally under 100ms to output version/help text.)

On PyCharm this pattern works fine so it may be an issue with a different IDE? `if False` is basically what `if TYPE_CHECKING` does at runtime but it just avoids the import.

A

Reply:
Can't comment on PyCharm, but I believe this will impact all contributors that use VsCode

Reply:
OK. I've restored to actual objects. I might bring Sphinx 6 forwards so that we can use `from __future__ import annotations` sooner than April.

A

Reply:
`5.x` has been updated.